### PR TITLE
Implement a loader for dyck based challenges and references.

### PR DESCRIPTION
**Pull request by Sid-Data-Universe**
Generating 3 batches in a notebook (using random_seed=0) resulted in:
```
('Complete the rest of the sequence, making sure that the parentheses are closed properly. Input: < > ( ) { } { [ ] < > { < > } { } } [ ] ( ) < < > ( ) {', '} >')
('Complete the rest of the sequence, making sure that the parentheses are closed properly. Input: [ [ ( < > ) {', '} ] ]')
('Complete the rest of the sequence, making sure that the parentheses are closed properly. Input: < > [ ] ( ( ) { [ < > ] < [ < > ] [ ] > {', '} } )')
```

### CODE REVIEW

File: finetune/datasets/generated/dyck_loader.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            # Get index of all the close characters.
+            close_chars = {pair[1] for pair in DYCK_CHARACTER_PAIRS}
+            close_char_indexes = []
+            for index, char in enumerate(complete_dyck):
+                if char in close_chars:
+                    close_char_indexes.append(index)
+
+            chosen_indexes = random.sample(close_char_indexes, reference_length)
+            # Sort the chosen indexes so the reference answer has them in the right order.
+            chosen_indexes.sort()
```
Comment (Sid-Data-Universe) [inline finetune/datasets/generated/dyck_loader.py:116]: Note the sorting doesn't actually matter as we enumerate in order below.

However it is also not clear to my that this is ever deterministic.

Both ( < ) > and ( < < ) are valid dyck words so given ( < it is not clear what the reference should be.

Comment (Sid-Data-Universe) [reply finetune/datasets/generated/dyck_loader.py]: Resolved. For a Dyck language with multiple delimiters a stack approach is used.

File: finetune/datasets/generated/dyck_loader.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    """
+    open_count = 0
+    close_count = 0
+    unused_close_characters_stack = []
+    dyck_word = ""
+
+    # Generate the word.
+    for _ in range(pair_count * 2):
+        # Randomly choose to open or close unless we have nothing open or already have every pair opened.
+        if open_count - close_count == 0 or (
```
Comment (Sid-Data-Universe) [inline finetune/datasets/generated/dyck_loader.py:71]: Remove the close count and just check if close stack is empty

**Review by RusticLuftig — DISMISSED** [review]

**Review by RusticLuftig — APPROVED** [review]

File: finetune/datasets/generated/dyck_loader.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+        for _ in range(samples):
+            # Decide the length the reference should be.
+            reference_length = random.randint(min_length_answer, max_length_answer)
+            # We need at least as many pairs as the reference length or the min length.
+            adjusted_min_length_pairs = max(reference_length, min_length_pairs)
+            # Decide total number of pairs.
+            pair_count = random.randint(adjusted_min_length_pairs, max_length_pairs)
+
+            # Generate a dyck of the minimum pair length.
```
Comment (RusticLuftig) [inline finetune/datasets/generated/dyck_loader.py:124]: nit: comment doesn't align with the code

**Review by RusticLuftig — APPROVED** [review]

File: finetune/datasets/generated/dyck_loader.py (hunk 3 above)
Comment (Sid-Data-Universe) [reply finetune/datasets/generated/dyck_loader.py]: Fixed. thanks!

**Review by RusticLuftig — APPROVED** [review]

**Pull request merged**